### PR TITLE
For #6324 Allow customization of display url in toolbar

### PR DESCRIPTION
**Pull request by kglazko**
---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by kglazko** [comment]
The scope of this PR changed to just shortening prefixes (https/http/www) and a separate PR will be made for centering/trimming long URLs.

File: components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt [hunk 1]
```diff
@@ -101,4 +101,44 @@ object URLStringUtils {
...
+             * If there is no http/https or www prefixes then the method will return the original Url
+             * for the displayUrl parameter in the DisplayUrl object.
+             */
+            fun from(url: CharSequence): DisplayUrl = DisplayUrl(url, maybeStripUrlPrefix(url))
+
+            private const val HTTP = "http://"
+            private const val HTTPS = "https://"
+            private const val WWW = "www."
+
+            private fun maybeStripUrlPrefix(url: CharSequence): CharSequence {
```
Comment (csadilek) [inline components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt]: nit: `maybeStripProtocol` or `maybeStripUrlScheme`.

File: components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt [hunk 2]
```diff
@@ -101,4 +101,44 @@ object URLStringUtils {
...
+                var noPrefixUrl = url
+                if (url.toString().startsWith(HTTPS)) {
+                    noPrefixUrl = maybeStripUrlWWW(url.toString().replaceFirst(HTTPS, ""))
+                } else if (url.toString().startsWith(HTTP)) {
+                    noPrefixUrl = maybeStripUrlWWW(url.toString().replaceFirst(HTTP, ""))
+                }
+                return noPrefixUrl
+            }
+
+            private fun maybeStripUrlWWW(url: CharSequence): CharSequence {
```
Comment (csadilek) [inline components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt]: nit `maybeStripSubdomain`?

File: components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt [hunk 3]
```diff
@@ -101,4 +101,44 @@ object URLStringUtils {
     // complain that this value is not one of the predefined enums that are allowed.
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     internal var flags = 0
+
+    /**
+     * This class is used to hold an original and shortened url
+     * The shortened url called displayUrl is a Url that is stripped
+     * of its prefixes.
+     */
+    data class DisplayUrl(val originalUrl: CharSequence, val displayUrl: CharSequence) {
```
[the remark this answers is not in the document]
Comment (csadilek) [reply components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt]: So, I think this can all turn into a single `toDisplayUrl(val originalUrl...)` if we don't need the state.

File: components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt [hunk 4]
```diff
@@ -160,4 +160,34 @@ class URLStringUtilsTest {
         assertTrue(isSearchTerm("www.c-c-  "))
         assertTrue(isSearchTerm("3-3 "))
     }
+
+    @Test
+    fun urlPrefixHttps() {
```
Comment (csadilek) [inline components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt]: Thanks for adding these test cases. If we apply the change above you can fold them into a `toDisplayUrl` test, or keep separate but then I would adjust naming to indicate what the test does e.g. `stripUrlScheme` etc.

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt [hunk 5]
```diff
@@ -439,10 +440,13 @@ class DisplayToolbar internal constructor(
     /**
      * Updates the URL to be displayed.
      */
+    private var _url: URLStringUtils.DisplayUrl? = null
```
[the remark this answers is not in the document]
Comment (csadilek) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/display/DisplayToolbar.kt]: And since we don't need hold state anymore in `DisplayUrl` this would become: 

```
 /**
     * Updates the URL to be displayed.
     */
    internal var url: CharSequence = ""
        set(value) {
            field = value
            views.origin.url = URLStringUtils.toDisplayUrl(value)
            updateIndicatorVisibility()
        }
```

File: components/browser/toolbar/src/test/java/mozilla/components/browser/toolbar/display/DisplayToolbarTest.kt [hunk 6]
```diff
@@ -698,4 +698,14 @@ class DisplayToolbarTest {
         menuView.dismissMenu()
         assertTrue(wasDismissed)
     }
+
+    @Test
+    fun `url formatter used if provided`() {
+        val (_, displayToolbar) = createDisplayToolbar()
+        displayToolbar.url = "https://mozilla.org"
+        assertEquals(displayToolbar.url, displayToolbar.views.origin.url)
```
Comment (csadilek) [inline components/browser/toolbar/src/test/java/mozilla/components/browser/toolbar/display/DisplayToolbarTest.kt]: nit: a newline here would make this easier to read. 2 different cases.

File: components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt [hunk 7]
```diff
@@ -101,4 +101,33 @@ object URLStringUtils {
     // complain that this value is not one of the predefined enums that are allowed.
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     internal var flags = 0
+
+    private const val HTTP = "http://"
+    private const val HTTPS = "https://"
+    private const val WWW = "www."
+
+    /**
+     * This method uses the un-shortened URL to generated a shortened URL
```
Comment (csadilek) [inline components/support/utils/src/main/java/mozilla/components/support/utils/URLStringUtils.kt]: nit: to generate, not generated.

Maybe let's rephrase this to:

Generates a shorter version of the provided URL for display purpose by stripping it of...

File: components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt [hunk 8]
```diff
@@ -160,4 +160,34 @@ class URLStringUtilsTest {
...
+    }
+
+    @Test
+    fun stripUrlSchemeUrlWithHttp() {
+        val testDisplayUrl = URLStringUtils.toDisplayUrl("http://mozilla.com")
+        assertEquals("mozilla.com", testDisplayUrl)
+    }
+
+    @Test
+    fun stripUrlSubdomainUrlWithHttpsWWW() {
```
Comment (csadilek) [inline components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt]: nit: let's remove the WWW from the name (same as test below.)

File: components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt [hunk 9]
```diff
@@ -160,4 +160,34 @@ class URLStringUtilsTest {
...
+    }
+
+    @Test
+    fun stripUrlSubdomainUrlWithHttp() {
+        val testDisplayUrl = URLStringUtils.toDisplayUrl("http://www.mozilla.com")
+        assertEquals("mozilla.com", testDisplayUrl)
+    }
+
+    @Test
+    fun stripUrlSchemeAndSubdomainUrlOther() {
```
Comment (csadilek) [inline components/support/utils/src/test/java/mozilla/components/support/utils/URLStringUtilsTest.kt]: nit: `stripUrlSchemeAndSubdomainUrlNoMatch`
